### DISCUSSION TITLE
Split video from scenes file?

### DISCUSSION THREAD

**Discussion by leingang** [Breakthrough/PySceneDetect#275] [Q&A]
It seems like I use this pattern a lot: 

- `scenedetect -i file.mp4 -s stats.csv detect-threshold list-scenes`
- Inspect scenes file to see if it's reasonable
- If not, inspect stats file to find a better threshold
- `scenedetect -i file.mp4 detect-threshold [-t nn] split-video`, using a different threshold than default if necessary.

In the case when the scenes described by the first pass are good, it would be great to split the video from the scenes file instead of doing another full scan. Is there a built-in way to do that?

----

Thinking about this a little more, if the functionality isn't there yet, there could be another "detector" that reads scenes from a text file. Naturally, it should be able to read the CSV file generated by `list-scenes`.

I'd also like to be able to quickly adjust the scenes manually. If line n-1 describes a scene from times A to B, line n from B to C, and line n+1 from C to D, and I delete line n, I would like the scene to go from A to C and C to D. But I can see another use case where deleting line n would split scenes from A to B and C to D (skipping over B to C completely).

I'll poke around in the source to see how this might work.

**Comment by Breakthrough** [comment]
There isn't a built-in way of doing this, but it's certainly feasible to add this to PySceneDetect.  It might be easier to just add a new argument to `split-video` for this purpose:  (Edit: your original suggestion to add another command for this seems like a better long-term solution)

    scenedetect -i video.mp4 split-video -i scenes.csv

Feel free to submit a feature request for this in the issue tracker.

**Reply by Breakthrough** [reply]
I think I see your point now about why this should be a "detector", given then it would also work seamlessly with other commands (e.g. `save-images`).  That approach also seems reasonable, maybe could add a `load-scenes` command, e.g.:

    scenedetect -i video.mp4 load-scenes -i scenelist.csv split-video

And then just make it mutually exclusive with actual detectors.

**Reply by Breakthrough** [reply]
Also the only important thing is having a start time and a duration/length of any format, so more compact forms could also be allowed so long as the column names are standardized.  This would make adjusting scenes manually easier, e.g. allow any of these as inputs:

```
Start Timecode,End Timecode
1,49,00:00:02.002,2.002,90,00:00:03.754,3.754,42,00:00:01.752,1.752
2,91,00:00:03.754,3.754,193,00:00:08.050,8.050,103,00:00:04.296,4.296
```

```
Start Frame,End Frame
49,90
91,194
```

Then you can just use any CSV editor to remove the redundant columns manually to make editing easier.  Thoughts?

**Reply by leingang** [reply]
Yes! Using a `SceneDetector` to read scenes from an external file seems to fit both the use case architecture (works with split-video, save-images, etc., and the application framework (A `SceneDetector` produces a list of scenes) well. 

And yes, using the same column names as in the output scenes CSV file would make it easier to adjust scenes. If I want to delete a cut and stitch together two scenes, I would:

1. open and edit the CSV file to remove the redundant columns
2. change the end point of a scene to the end point of the following scene
3. delete the row corresponding to the following scene
4. save and feed the new CSV file back into scenedetect.

If I want to add a cut point and insert a scene, I'd need to:

1. open and edit the CSV file to remove the redundant columns
2. add a row after the scene I want to split
3. add a start point which is the new cut point
4. add an end point which is the end point of the scene to be split
5. change the end point of the scene to be split to be the cut point.
6. save and feed the new CSV file back into scenedetect.

**Reply by Breakthrough** [reply]
That sounds like a good approach.  Also now that config files are a thing, we could also add an option in the config file to specify the exact columns you want when running `list-scenes` (which would save having to remove the redundant columns each time).

**Comment by Breakthrough** [comment]
Just realized this is already covered by #235, I'll link to this thread there as there's some good ideas to follow up with here.

**Comment by wjs018** [comment]
@leingang 

Have an early version of this feature implemented in #294. There are some current limitation of the api that prevent all of the discussed ideas in this thread from being implemented, but if you have any feedback on the usage of the `load-scenes` command, let me know.
